src/Provider/AbstractProvider.php

```diff
@@ -98,7 +108,15 @@ public function __construct($options = [], HttpAdapterInterface $httpClient = nu
             }
         }
 
-        $this->setHttpClient($httpClient ?: new CurlHttpAdapter());
+        if (empty($collaborators['httpClient'])) {
+            $collaborators['httpClient'] = new CurlHttpAdapter();
+        }
+        $this->setHttpClient($collaborators['httpClient']);
+
+        if (empty($collaborators['randomFactory'])) {
+            $collaborators['randomFactory'] = new RandomFactory();
+        }
+        $this->setRandomFactory($collaborators['randomFactory']);
```

Comment:
@ramsey we clearly have a misunderstanding. The RandomLib factory allows users to inject their own instances of Source/Mixer into the factory using [registerSource](https://github.com/ircmaxell/RandomLib/blob/338d2f594f7148983805ad56f47de7e47d84b2b9/lib/RandomLib/Factory.php#L157) and `registerMixer`. The generator relies on Source/Mixer to provide the actual values, the generator simply acts as a manager.